### ISSUE TITLE
Experience very slow connection between snowflake

### ISSUE THREAD

**Issue by billstark** [hashicorp/vault-plugin-database-snowflake#13]
Hi there, currently we are experiencing some problems creating database creds with snowflake on Vault.

Vault Version: 1.9.3

## How we created the database engine

```bash
# add config
vault write database/config/sg-datadev-snowflake \
    plugin_name=snowflake-database-plugin \
    allowed_roles="sg-datadev-snowflake-loader-rw" \
    connection_url="{{username}}:{{password}}@xxxxxxx.xxxxxxxxx.azure" \
    username="vault_user" \
    password="vault_user"

# add secret
vault write database/roles/sg-datadev-snowflake-loader-rw \
    db_name=sg-datadev-snowflake \
    creation_statements="USE ROLE USERADMIN; CREATE USER {{name}} PASSWORD = '{{password}}' DAYS_TO_EXPIRY = {{expiration}} DEFAULT_ROLE=LOADER DEFAULT_WAREHOUSE=LOAD_WH; USE ROLE SECURITYADMIN; GRANT ROLE LOADER TO USER {{name}}; ALTER USER {{name}} SET NETWORK_POLICY = allow_only_dev_vpc;" \
    default_ttl="10m" max_ttl="10m"
```

The problem was that sometimes 
<img width="2106" alt="WX20220414-183227@2x" src="https://user-images.githubusercontent.com/16464626/163374861-69fcc768-35ba-446b-85d6-75237bb11057.png">
we are not able to get secrets and when it succeeds, generally it takes about more than 30 seconds. I have attached some of queries below. Would you please help us check about this issue? Thanks!

**Comment by calvn** [comment]
Hi @billstark, thanks for sharing the config and sample output of the query times. I have a few follow-up questions regarding this:

1. Does this occur on all credentials creation requests, or does it occur sporadically? If it's the latter, how often does it occur, and whether it correlates with other queries that the DB might be processing at that time.
2. Do you have issues if users were created manually and separately?
3. Can you try on a separate DB if that's available?

**Comment by billstark** [comment]
Hi @calvn , regarding the questions
1. This occurs for all credential creation queries.
2. Nope, we do not have issue to create users on Snowflake UI
3. In our production environment, which has lower version `1.7.0` does not have this issue

**Comment by swenson** [comment]
Hmm. I'm definitely seeing consistently slower behavior by a factor of ~4× on Vault `1.10.0`:

![image](https://user-images.githubusercontent.com/33404/164333673-9b8c7773-292c-4113-9952-ecfc2218041d.png)

As opposed to Vault `1.7.0`:

![image](https://user-images.githubusercontent.com/33404/164333770-f9b0169c-a18b-46fd-868f-fa7a0be4eafb.png)

Is this the slowdown you're referring to @billstark? The 1+ second creation times (for `1.10.0`) as opposed to <1 seconds (for `1.7.0`)?

I'll see if I can bisect when this slowdown started.

**Comment by swenson** [comment]
I believe I have identified the issue, assuming the `CREATE USER` slowdown is the issue you're referring to, @billstark.

It was introduced in Vault 1.8.2 when the underlying driver https://github.com/snowflakedb/gosnowflake was updated to v1.6.1.

I bisected it down further to be a slowdown introduced in `gosnowflake` 1.4.1.

I have a fix that I confirmed reduces the `CREATE USER` times back down to where they were in Vault <= 1.8.1: https://github.com/snowflakedb/gosnowflake/pull/574

**Comment by billstark** [comment]
Hi @swenson thanks a lot for your quick action! Yes, I think you are right that was the slowdown that I am referring to. But in this case, after this problem has been fixed, may I know how would we be able to adapt to this fix?

**Comment by swenson** [comment]
@billstark It will depend. If the Snowflake team are quick to merge the PR or find another fix, I'll try to be quick on our side to test and update the dependency on ours.

If it takes more than a few days, I'll consult with the rest of the team and see if we will want to temporarily form `gosnowflake` with this performance fix, and update our dependency temporarily.

**Comment by swenson** [comment]
Okay, we have a fix that has been merged in the latest `main` branch of the plugin, which should be released in the next major release of Vault (1.11).
